### PR TITLE
Vulkan revision 2

### PR DESCRIPTION
**Pull request by haasn**
This improves on the major shortcomings of the ver1 design and enables full use of async transfers, async compute, multi-command queueing, event-based image barriers, safe FBO invalidation, better implicit renderpass layout transitions and more (tm).

It turns out that the VkSemaphore deadlock issue was an nvidia driver bug. This approach works after all, and scales well - especially now that I've refactored it to avoid using VkSemaphores for intra-frame dependencies as much as possible. There is now virtually no performance drop as a result of this refactor even without the fancy async transfer stuff. (And with async transfer on top, the performance goes through the roof again)

I probably won't actually merge this until the fixes for the nvidia deadlock issue make it into an actual driver release, but it would be good to get some testing (on other platforms) and code review out of the way.

NOTE: This is still not entirely optimal. Improvement avenues:

1. Try and avoid over-abusing CONCURRENT sharing mode (hard)
2. Try and avoid using VkEvents in trivial scenarios (instead use pipeline barriers). (trivial)
3. Try recreating VkSemaphores and re-using the VkEvents instead for already-submitted signals. (trivial)
4. Try improving the logic for when to split the command buffers to maximize performance. (easy-ish but annoying)
5. Try updating buffers from different queues depending on their type. (medium)
6. Try and use LOAD_OP_CLEAR to clear color attachments instead of doing the stupid 1x1 blit / vkClear thing. (easy-medium)
7. Allow recording timestamps from the COMPUTE queue as well as the GRAPHICS queue, to avoid over-submitting tiny command buffers. This should especially help for async compute. (easy)

And probably more that I've forgotten. Let the hate commence!

### CODE REVIEW

**Comment by haasn** [comment]
> Try and use LOAD_OP_CLEAR to clear color attachments instead of doing the stupid 1x1 blit / vkClear thing. (easy-medium)

Gave this a try but it was slower on nvidia. And I actually have no idea why, because even just uncommenting the `clear()` call made it slower; probably because it drops the implicit invalidation on the framebuffer, but despite this it was *still* slower even if I force-invalidated in `pass_draw_to_screen`.

Still needs testing on AMD I suppose.

**Comment by haasn** [comment]
The test feedback so far has been very positive for windows; and also works great on nvidia, and the official AMD drivers. It also fixes some issues people were having, and makes it faster and more responsive.

RADV/mesa is the only platform that doesn't seem to like it; but I don't think that will hold us back. (Maybe merging it into master will give mesa some motivation to fix the issues?)

**Comment by haasn** [comment]
Rebased. I've also added an option to disable async transfer/compute, mainly for debugging purposes, but also because they seem to be unstable on some drivers (async compute in particular).

**Comment by haasn** [comment]
Nvidia driver version is out and should include the bug fixes for this. With async compute disabled by default it also avoids crashes on several other drivers.

I'd be happy to merge it in this form. @wm4 ?

**Comment by haasn** [comment]
Clang just pointed out a fairly glaring issue in the sharing mode initialization for allocated buffers. In theory, that *might* have explained some of the issues with async transfer/compute on devices where it led to issues.

(Although probably the only affected buffers are texture upload buffers)

**Comment by haasn** [comment]
> (Although probably the only affected buffers are texture upload buffers)

Actually, I'm not even sure if they're affected. I think the only case that was affected were updates for uniform buffers, because UBO buffers are always updated from the graphics queue (for stupid reasons) even when using them with compute shaders.

So that actually makes sense when you think about the async compute stuff. Hopefully this fixed it? If users could try with the above fix and --async-compute again I'd be very grateful.

**Comment by haasn** [comment]
Found and fixed some bugs relating to image synchronization, thanks to a vulkan validation layers update. It's possible that this fixes the crashing on some devices. Testing desired.

**Comment by aufkrawall** [comment]
Does --vulkan-async-compute work for anyone? With Pascal and 387 drivers, I get "device timeout" errors on Windows and it semi-freezes the whole system on Linux.
(But I guess there won't be any performance wins anyway with Pascal if it's compareable to games.)

**Comment by haasn** [comment]
@aufkrawall Can you test with 4a19ae5?

**Comment by aufkrawall** [comment]
It's fixed now (tested on Linux). It doesn't have an influence on the reported frame timings though, GPU usage reported by the driver is also roughly the same.
Do GCN GPUs already perform better with it?

**Comment by haasn** [comment]
@aufkrawall I don't know of any GCN results, but I've heard reports from a user that said async-compute bumped FPS from 520 to 550 on a 1080 Ti.

Apparently the latest round of fixes, plus the new nvidia driver version, means -rev2 will no longer crash or deadlock on windows nvidia. (I still have one swapchain hang on linux nvidia but I suspect a driver bug here as well, since it matches the pattern of the other swapchain-related issues)

Don't know if any AMD Windows users have tested this version. Mesa still doesn't work.

**Comment by Artoriuz** [comment]
I have a RX470 and I'm currently using Windows. 

The last time I tried to test the difference on a 540p AVC file (decodes fast) with gpu-hq + ravu-r4 + adaptive-sharpen (just to increase the GPU load), with --vulkan-async-compute and --vulkan--queue-count=3 I got around 490 fps. Without --vulkan-async-compute and --vulkan--queue-count=3 I got around 515 fps. 

I could time how long it takes to reach 1min mark or something in both cases, I feel that with async-compute turned on the minimums are higher, but I could be wrong.

**Comment by AstralStorm** [comment]
It is actually broken on Linux and amdgpu-pro when running color conversion. (At least yuv420.)
Valgrind suggests an overrun in mpv code, but couldn't pinpoint it despite full debug info. All thanks to amdgpu-pro not having debug symbols. RGB works more or less fine on that ICD.
(Yes, still kills radv.)

**Comment by haasn** [comment]
@AstralStorm: What does ‘broken’ mean in this context?

**Comment by sl1pkn07** [comment]
please rebase, patch failed when apply the PR in master (wscript)

**Comment by AstralStorm** [comment]
@haasn Crashing on surface swap regardless of sync or async or compute mode.

**Comment by haasn** [comment]
@AstralStorm: backtrace? log?

**Comment by AstralStorm** [comment]
Backtrace is useless due to lack of symbols. (Cannot trace back from ICD.) It crashes on the second real frame swap. I will provide the log soon.

**Comment by haasn** [comment]
@sl1pkn07 rebased; somebody decided to make the same change to mpv master as one of the commits in this PR but did it slightly differently, which was the reason for the conflict...

Untested (ffmpeg master still doesn't work with mpv)

**Comment by haasn** [comment]
@AstralStorm can you reproduce it from inside gdb, and maybe step through the function `submit_frame` from vulkan/context.c? (or just break on ra_vk_submit and go from there)

**Comment by sl1pkn07** [comment]
Hi

seems this https://github.com/mpv-player/mpv/pull/4933/commits/d98c538b914ff0ef09ae31bda5a1bfcaf29f0384, the first hunk, is applied

or patch the PR see it

```
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file DOCS/man/options.rst.rej
patching file video/out/vulkan/context.c
patching file video/out/vulkan/ra_vk.c
```

also conflicts in video/out/gpu/ra.h

the patch is applied in clean clone

greetings

**Comment by haasn** [comment]
Rebased again. Still no idea if it works or not.

**Comment by DeadSix27** [comment]
@haasn What are the supported chroma scalers in vulkan? All?
I have this odd issue when using `cscale=ewa_lanczossoft` and `gpu-api=vulkan`: 

image

Am I doing something wrong? (Also, this does not happen when using `cscale=haasnsoft`)
(I'm running a AMD 390x)

**Comment by haasn** [comment]
@DeadSix27 Sounds like a synchronization issue. Can you post a log with rev2?

**Comment by CrisBRM** [comment]
Here's some mesa amdgpu feedback (280x): it works just fine. Tried a bunch of different shaders. Performance is pretty much the same between this one and rev1, if only ever so slightly 'slower'.

**Comment by DeadSix27** [comment]
@haasn log-file: https://0x0.st/siOG.txt (Any specific debug parameter I should use? (I just used logfile))

**Comment by haasn** [comment]
That seems like a rev1 log. Are you sure you've applied the rev2 patches correctly? Your problem is also indiciative of rev1, since the synchronization required doesn't exist until rev2.

**Comment by DeadSix27** [comment]
@haasn I was pretty sure up until this point.. Let me do it all manually maybe I derped or uploaded the wrong output log.

**Comment by DeadSix27** [comment]
@haasn PR was not applied correctly, you're correct, it works flawlessly with your changes, but I used  Just rev1 has that sync issue.
Why it wasn't applying your PR? Idk I used: `git fetch origin pull/4933/head:master`.. which I thought would work perfectly.. that's exactly how github tells one to do it: `https://help.github.com/articles/checking-out-pull-requests-locally/` .. I'll use the auto generated patch from now on.

**Comment by haasn** [comment]
@DeadSix27 `git fetch` only fetches heads. It doesn't merge them. You probably meant `git pull`, which is fetch+merge in one.

**Comment by DeadSix27** [comment]
@haasn Indeed, I looked back into it and it requires a new branch and merging to do what I thought it would do.

**Comment by DeadSix27** [comment]
@haasn One more thing, is it possible to not use exclusive fullscreen with Vulkan? In OpenGL that was done via dwmflush no? Not sure, I never used it since non-exclusive FS was default behavior.

**Comment by haasn** [comment]
> One more thing, is it possible to not use exclusive fullscreen with Vulkan?

It's possible to explicitly take over a display device using the `VK_KHR_display_*` family of extensions; thus removing it from the normal compositor. (e.g. for HMDs)

Seems like all you want is for us to *not* explicitly do this, which is indeed what we (don't) do. OTOH, if you still get something like “exclusive” fullscreen mode (or fullscreen unredirection) despite this, there's nothing that can be done except fiddling with driver settings, since afaik there's no vulkan extension to control fullscreen behavior like that.

**Comment by haasn** [comment]
> But ye rv2 runs and looks just fine.

So can you confirm that -rev2 works fine with amdgpu now?

**Comment by DeadSix27** [comment]
@haasn Ye sadly it still goes exclusive, GL and Angle don't. Also doubt AMD has a setting for this.
Apart from that i can confirm it works without any issue on Win10 with an AMD 390x GPU.

**Comment by CrisBRM** [comment]
I can confirm it works just fine with amdgpu.

**Comment by AstralStorm** [comment]
Hanged here (incl. GPU) with both radv and amdgpu-pro on actual amdgpu, 4.14.2 kernel, R9 Nano.

**Comment by Artoriuz** [comment]
Any reason this just can't get merged already? People have been used builds with it for ages by now and all alarming issues ignoring window-screenshots seem to have been fixed.

**Comment by aufkrawall** [comment]
Hopefully this will be merged before next release. :(

**Comment by haasn** [comment]
At least one of the radeon developers is looking into the issue with amdgpu and some generations of AMD GPUs; but for the meantime, disabling the use of async transfer avoids triggering the stall (which happens when attempting a buffer->image copy from the compute queue).

**Comment by DeadSix27** [comment]
@haasn Is he a developer **from** radeon ? If so, leave a comment about the weird exclusive fullscreen issue.

**Comment by aufkrawall** [comment]
Are you using Windows 10 Creators Update 2? It should by default force every fullscreen application into "direct borderless" mode (which should be what you want), unless you disabled that in the Windows compatibility options for mpv.exe.
I wouldn't bet AMD changes the behavior when this is basically already done by a recent Win 10 version.

**Comment by DeadSix27** [comment]
@aufkrawall You mean this?: 
image I tried it on/off, no change.

**Comment by jeeb** [comment]
@Artoriuz I did poke @haasn about this a couple of days ago and it seemed like he wanted to wait for the amdgpu fixes, but I might have understood him wrong. As long as this has close to zero effect on anything else I'd just take this in since vulkan has more or less been experimental in mpv so far as well.

**Comment by aufkrawall** [comment]
For me it worked absolutely reliably so far on Nvidia Pascal (Windows 10/Arch Linux) and Intel Skylake (Arch Linux) so far, except of that async compute segfault which haasn fixed. :)

**Comment by haasn** [comment]
@jeeb I'd take it in as-is since we've hit a stalemate in the bug fixing process. I can't find any more issues (apart from the ones I discovered while re-analysing/rewriting this code for libplacebo); so anything remaining would need investigation from driver devs or improvements to the debugging layers.

One thing I'm sort of wishing we could do is make the use of `VkEvents` (which is the major addition of this patch) optional as well, but since you can also work around the issues by using `--vulkan-async-transfer=no` on affected devices it's not that critical.

**Comment by kevmitch** [comment]
Here are my testing results:

Intel Ivy Bridge, mesa 17.2.5, linux 4.7.10. Running `--no-config --deband=yes --video-sync=display-resample`.  Not sure if we care about this platform, since with both revisions, the driver prints `WARNING: Ivy Bridge Vulkan support is incomplete` and also
```
../../../src/intel/vulkan/gen7_cmd_buffer.c:273: FINISHME: Implement events on gen7
../../../src/intel/vulkan/gen7_cmd_buffer.c:297: FINISHME: Implement events on gen7
```

but it actually seems to mostly work in both rev1 and rev2 other than high CPU usage of about 80% of one core.

The average redraw time (as indicated by stats.lua) is increased with rev2 although the fresh draw time is slightly lowered. 

vulkan_rev1: 13 ms draw / 3 ms redraw ([log](http://sprunge.us/bMfG), [stats](http://sprunge.us/YWOJ))
vulkan_rev2: 12 ms draw / 11 ms redraw ([log](http://sprunge.us/JBXf), [stats](http://sprunge.us/JOIQ))

I also see lots of `[vo/gpu] Error: texture could not be created.` that were not previously there in rev1. Interestingly this no longer appears if I switch my refresh rate down to 24fps. This also seems to go away if i turn off display sync or debanding.

None of this seems to be affected by the various permutations of `--vulkan-async-transfer` and `--vulkan-async-compute` .

**Comment by haasn** [comment]
@kevmitch The timers are unreliable with rev2. Can you try using [`--profile=bench](https://haasn.xyz/posts/2017-10-05-how-to-benchmark-mpvs-raw-throughput.html)?

**Comment by kevmitch** [comment]
## base
rev2 is 3% faster

`--no-config --gpu-api=vulkan --audio=no --untimed --video-sync=display-desync --vulkan-swap-mode=immediate --opengl-swapinterval=0 --osd-msg1="FPS: \${estimated-display-fps}" --fs`

vulkan-rev2: 370fps [log](http://termbin.com/iikp), [stats](http://termbin.com/pmge)
vulkan-rev1: 360fps [log](http://termbin.com/jp5a), [stats](http://termbin.com/hlfo)

## deband
rev2 is 20% slower

`--no-config --gpu-api=vulkan --deband --audio=no --untimed --video-sync=display-desync --vulkan-swap-mode=immediate --opengl-swapinterval=0 --osd-msg1="FPS: \${estimated-display-fps}" --fs`

vulkan-rev2: 180fps [log](http://termbin.com/yeay), [stats](http://termbin.com/2fos)
vulkan-rev1: 220fps [log](http://termbin.com/zyt8), [stats](http://termbin.com/273o)

----
Again, `vulkan-async-compute` and `vulkan-async-transfer` seem to have no measureable effect.
I wanted to compare with opengl, but it appears `--opengl-swapinterval=0` doesn't actually work in my case as it's stuck to my oddball display fps of 59.68.

**Comment by jeeb** [comment]
OK. Took a general look at this whole thing and then tried to focus on the effects it could have on the other renderers since there was limited touching of generic files included.

* Linux+wayland+opengl: builds & works on a X230 (similar Ivy Bridge to @kevmitch). Vulkan builds but as I lack shaderc there's nothing to generate SPIR-V for mesa :) .
* Windows+gpu contexts {d3d11,angle,win} with and without dxva2-copy or d3d11va hw decoders: all seems good. Didn't yet build Vulkan as it seems like more patching is required for the Kronos stuff, but will do that after I get some sleep.
* Didn't yet test android but highly likely to have not been changed.

Otherwise the biggest thought I got after looking at the code was:
> Is `bool discard` is good enough of a name, or should it be `bool discard_texture` or something? Additionally, the value gets right away set to a variable called `invalidate_target`.

Thus, I have very minimal concerns (although I'm still interested in the logic behind the parameter grouping in those functions where the new parameters were added). Additionally ran some [static analysis](https://kuroko.fushizen.eu/static_analysis/mpv/scan-build-2017-12-15-012641-31726-1/) on the current master with vulkan_r2 rebased on top. There are some warnings from vulkan-related files but not sure if those have anything to do with code changed in this PR.

**tl;dr** Looking pretty good as far as I'm generally concerned.

**Comment by kevmitch** [comment]
> Which line throws the error
```
    vo/gpu: Error: texture could not be created.

Thread 11 "mpv/vo" hit Breakpoint 1, ra_tex_resize (ra=0x7fffbc4d0780, log=0x5555574ce020, 
    tex=0x7fffbc5626d0, w=650, h=352, fmt=0x7fffbc4d13e0) at ../video/out/gpu/utils.c:207
207             mp_err(log, "Error: texture could not be created.\n");
(gdb) bt
#0  ra_tex_resize (ra=0x7fffbc4d0780, log=0x5555574ce020, tex=0x7fffbc5626d0, w=650, h=352, 
    fmt=0x7fffbc4d13e0) at ../video/out/gpu/utils.c:207
#1  0x00005555557a6886 in gl_video_render_frame (p=0x7fffbc562270, frame=0x7fffbc6a7b50, fbo=...)
    at ../video/out/gpu/video.c:3067
#2  0x00005555557b8575 in draw_frame (vo=0x5555574c5fe0, frame=0x7fffbc6a7b50) at ../video/out/vo_gpu.c:88
#3  0x00005555557b55f1 in render_frame (vo=0x5555574c5fe0) at ../video/out/vo.c:866
#4  vo_thread (ptr=0x5555574c5fe0) at ../video/out/vo.c:970
#5  0x00007fffec7b4519 in start_thread (arg=0x7fffd8ff9700) at pthread_create.c:456
#6  0x00007ffff2deda4f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

ultimately because `video/out/vulkan/ra_vk.c:vk_tex_create:549`
```
    VkResult res = vkGetPhysicalDeviceImageFormatProperties(vk->physd,
            fmt->iformat, tex_vk->type, VK_IMAGE_TILING_OPTIMAL, usage, 0,
            &iprop);
    if (res == VK_ERROR_FORMAT_NOT_SUPPORTED) {
        return NULL;
```
where  `fmt->name == "bgra8"`

> Can you try using `--vulkan-debug` as well? 

```
Error parsing option vulkan-debug (option not found)
Setting commandline option --vulkan-debug= failed.
```
Are you talking about this: https://gpuopen.com/using-the-vulkan-validation-layers/ ?


> my gut feeling tells me we might be exhausting your VRAM?

how do I measure this?

**Comment by sl1pkn07** [comment]
>how do I measure this?

If you have nvidia, with nvidia-smi, or nvidia-settings

**Comment by haasn** [comment]
> `Error parsing option vulkan-debug (option not found)`

Sorry, the option name is `--gpu-debug` not `--vulkan-debug`. (Maybe this should be split up into vulkan and opengl?)

The stack trace helps. Can you paste the output of `vulkaninfo`?

**Comment by kevmitch** [comment]
[vulkaninfo](http://sprunge.us/Njeh)

rev2 `--gpu-debug`: [log](http://sprunge.us/VAJM), [stats](http://sprunge.us/gJSP)

> (Maybe this should be split up into vulkan and opengl?)

I think it's fine as one option. It's doing the same high level thing. Does `d3d11` have an equivalent?

**Comment by jeeb** [comment]
Just tested and `--gpu-debug` works with d3d11 as well :) First seeing the GPU leak check, too.

**Comment by haasn** [comment]
@kevmitch The issue in your case is the fact that in mpv we universally try and enable storage image support for our intermediate FBOs, but your device doesn't support storage images on the `bgra8` fbo format. The `bgra8` texture format is used for the cache FBO because it's the same as your swapchain format. This is different compared to rev2 due to commit 9789bac.

This also explain the performance loss - your redraws are not getting cached as a result, since it fails creating the cache FBO. This should not cause any issues apart from the performance loss and the spammed error message. (Maybe we remember a failure here?)

The proper solution is to expand `ra_format` in order to allow the `ra_tex_resize` code to know whether the format is storable or not. This is done in libplacebo, but backporting it to mpv will require some work (partly because of the awful OpenGL texture format semantics).

The next-best half-baked work-around would be to add a flag to ra_tex_resize indicating whether we want to use it as a storage image or not. (The cache FBO is never used as a storage image). But this would just work around your specific issue - it doesn't completely eliminate the bug, since it's still possible we might run into cases of trying to make storable FBOs with a non-storable `fbo-format` in general, even *if* we revert the commit 9789bac. The root problem here is the fact that mpv's RA can't express the circumstance where some FBO formats are storable and others are not. In fact, you can probably reproduce a similar bug with OpenGL by picking some obscure fbo-format.

**Comment by haasn** [comment]
Anyway, I wouldn't mind merging this without commit 9789bac and leaving that commit to a later date. That way this at least isn't a regression; i.e. the only bugs related to non-storable FBO formats are also bugs in master.

**Comment by jeeb** [comment]
If we know the limitations of the current setup I do not consider this a deal-breaker, given that my understanding is that something is planned to be done about it (which might of course be incorrect). Better to not shake the pack too much compared to what has already gone through a relative amount of testing.

@kevmitch , what's your opinion?

**Comment by kevmitch** [comment]
I reverted the commit and indeed got a slight improvement over rev1

rev2 deband without 9789bac: 225fps [log](http://termbin.com/yn4k), [stats](http://termbin.com/98bn)

There are also no more `texture could not be created` messages. However, with `--gpu-debug` I now see a steady stream of vulkan warnings that were not present before reverting 9789bac:

```
[vo/gpu/vulkan] vk [DS] 10: Source AccessMask 0 [None] must have required access bit 2048 [VK_ACCESS_TRANSFER_READ_BIT]  when layout is VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, unless the app has previously added a barrier for this transition. (obj 0x7f4f30773940 (VkCommandBuffer), loc 0xa93)
```

**Comment by kevmitch** [comment]
well shit

**Comment by airlied** [comment]
Just FYI I posted a radv patch to fix the compute bugs on it, should be in master and next stable release.

**Comment by aufkrawall** [comment]
It's indeed working nicely here with Linux 4.15, mesa-git and RX 560.
Also gpu-queue-count 3 + async compute works.

According to render stats, async compute gives a massive performance boost (~4.5ms vs. ~14.5ms). Can that be real?
I already forced GPU into maximum clocks because the automatic clock adjustment of amdgpu doesn't work correctly for mpv.

With or without AC, I can run higher settings than with OGL. Kudos, haasn (and David Airlie ;) )!

**Comment by aufkrawall** [comment]
Ok, I used the more reliable benchmark method described by haasn in his blog (also added vsync off for D3D11). I don't have the conrete numbers anymore, but I checked everything several times, so I think my conclusions should be really reliable for my testcase (scaling down 4k 60fps 8 bit and 4k 50fps 10 bit to 1440p).

- On Windows, Vulkan can have huge gains over D3D11 (~12%). This seems especially true with 10 bit content (much > than 12%).
- With the Windows driver, async compute can show small gains (~5%) with certain workloads (e.g. cscale=catmull_rom + dscale=ewa_lanczossharp). This is also true with AMD's amdvlk on Linux.
- RadV doesn't seem to show any gains by AC. All drivers show a small loss with AC on when the workload doesn't seem optimal (too little graphics queue load?).
- However, RadV without AC is also faster than OpenGL when there is enough compute shader load.
- With all drivers, a queue count of 2 instead of 1 is a bit beneficial, but mostly with AC when AC is beneficial.
- On Windows, Vulkan significantly improves performance with D3D11VA-copy over D3D11, but is not faster than D3D11 + native D3D11VA (I suppose it would be awesome with a D3D11VA interop for Vulkan)
- Unlike RadV, amdvlk on Linux Xorg seems to have broken Vsync, I also noticed this in vkquake

You btw. still need mesa-git, Mesa 17.3.2 still hangs system with RadV + mpv.

**Comment by wm4** [comment]
> On Windows, Vulkan significantly improves performance with D3D11VA-copy over D3D11, but is not faster than D3D11 + native D3D11VA (I suppose it would be awesome with a D3D11VA interop for Vulkan)

I wonder if it'd be worth to do the hardware copy into Vulkan DR memory...

**Comment by aufkrawall** [comment]
I could also do the same test on a Windows 10 machine with a GTX 1060, also here Vulkan was faster than D3D11.
At least with d3d11va-copy, which I had to use to avoid CPU limitation due to software decoding.
However, with Pascal Async Compute can severely hurt performance when using ewa_lanczos for both cscale and scale at the same time while it also doesn't show any gains otherwise.

**Comment by aufkrawall** [comment]
RadV now works with Mesa 17.3.3. There is tearing in Xorg windowed, but not fullscreen.
Luckily, Compton vsync seems to "fix" windowed, it looks absolutely smooth to me with interpolation.
